src/lib.rs

```diff
@@ -11,6 +11,678 @@
+                format!("Unable to connect to mqtt broker: {e:?}"),
+            )
+        })?;
+
+        Ok(Self {
+            mqtt_client: Arc::new(mqtt_cli),
+            topic_listener_map,
+            authority_name,
+            client_type,
+        })
+    }
+
+    /// Helper function that handles MQTT messages on reception.
+    ///
+    /// # Arguments
+    /// * `cli` - MQTT client that received the message.
+    /// * `message` - MQTT message received.
+    fn on_receive(
+        _cli: &AsyncClient,
+        message: Option<paho_mqtt::Message>,
+        topic_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    ) {
+        if let Some(msg) = message {
+            let topic = msg.topic();
+
+            // Get attributes from mqtt header.
+            let uattributes =
+                UPClientMqtt::get_uattributes_from_mqtt_properties(msg.properties()).unwrap();
+
+            let payload = msg.payload();
+            let upayload = payload.to_vec();
+
+            let umessage = UMessage {
+                attributes: Some(uattributes).into(),
+                payload: Some(upayload.into()),
+                ..Default::default()
+            };
+
+            // Create UMessage from UAttributes and UPayload.
+            let topic_map_read = block_on(topic_map.read());
+            let listeners = topic_map_read.get(topic);
+
+            if let Some(listeners) = listeners {
+                for listener in listeners {
+                    let umsg_clone = umessage.clone(); // need to clone outside of closure.
+                    block_on(listener.on_receive(umsg_clone));
+                }
+            }
+        }
+    }
+
+    /// Get the client id of the mqtt client.
+    fn get_client_id(&self) -> String {
+        self.mqtt_client.client_id()
+    }
+
+    /// Get the client indicator based on client type.
+    fn get_client_indicator(&self) -> String {
+        match self.client_type {
+            UPClientMqttType::Device => "d",
+            UPClientMqttType::Cloud => "c",
+        }
+        .to_string()
+    }
+
+    /// Send UMessage to mqtt topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Mqtt topic to send message to.
+    /// * `message` - UMessage to send.
+    /// * `attributes` - UAttributes to send with message.
+    async fn send_message(
+        &self,
+        topic: &str,
+        message: &UMessage,
+        attributes: &UAttributes,
+    ) -> Result<(), UStatus> {
+        let props = UPClientMqtt::create_mqtt_properties_from_uattributes(attributes)?;
+        let data = message.payload.clone().unwrap(); // TODO: Handle optional case and is this the best way to send data?
+
+        let msg = mqtt::MessageBuilder::new()
+            .topic(topic)
+            .properties(props)
+            .payload(data)
+            .qos(1)
+            .finalize();
+
+        info!("Sending message: {:?}", msg);
+
+        // UserProperty is not sent with the message..
```

Comment:
Is that good thing? Bad thing? An issue?

Reply:
oh lol, that was a reminder to myself while debugging. removed